TST use global_random_seed in sklearn/cluster/tests/test_birch.py

marenwestermann opened the pull request:
#### Reference Issues/PRs
towards #22827


#### What does this implement/fix? Explain your changes.
Uses the global_random_seed fixture in the module `tests/test_birch.py`. 

#### Any other comments?
`test_birch_predict`:
In order to make all tests pass I changed the `threshold` parameter in the Birch estimator from 1.0 to 1.1 (line 62). 

`test_transform_match_across_dtypes`: 
Here I changed the `threshold` parameter in the Birch estimator from the default (0.5) to 1.1 (line 202) in order to make all tests pass. 

`test_feature_names_out` and `test_subcluster_dtype`: 
I didn't make use of the global_random_seed fixture in these tests because from my understanding it is not needed.

Commit:
test_birch [all random seeds]
test_birch_predict
test_birch_n_clusters_long_int
test_transform_match_across_dtypes

ClaudioSalvatoreArcidiacono reviewed (changes requested):
Hey @marenwestermann, hope you don't mind if I chime in :).

I noticed that you missed 2 `random_state`, one at line `191` and another line `210`. Rather than that LGTM.

ClaudioSalvatoreArcidiacono commented:
> Hey @marenwestermann, hope you don't mind if I chime in :).
> 
> I noticed that you missed 2 `random_state`, one at line `191` and another line `210`. Rather than that LGTM.

Just read your description and noticed that this was intentional, sorry for shooting too fast.

I would suggest to still change it for consistency. It is arguable whether fixing `global_random_seed` is needed for other tests that you propose to change in your PR, see for example `test_birch_n_clusters_long_int`.

What is your view on it?

ogrisel committed:
Merge branch 'main' into test-birch

marenwestermann commented:
Thank you for your reviews! I'll update the PR.

ogrisel commented:
It also seems that `test_threshold` has not random seed set for its call to `make_blobs`. I think we should use `global_random_seed` for this one as well.

ogrisel commented:
Similarly, I have the feeling that `test_branching_factor` might depend on the random values in the data, so better be safe and seed it as well.

marenwestermann committed:
address comments

ogrisel commented:
Thanks for the update. I am doing another pass on this file and there are more tests that would benefit from setting the `random_state` parameter of there data generating process using `global_random_seed`. Basically, all the tests that make assertions on values in fitted attributes such as `subclusters_` or `labels_` depend on the values in the training set (as opposed to just its shape, dtype or constructor params) and as such should be parametrized with `global_random_seed` to make sure that those assertions hold for various values of the random seed while at the same time not leave `random_state` to its default `None` value to make them run deterministically and avoid rare random failures in our test suite.

marenwestermann committed:
use global_random_seed for tests that make assertions on values in fitted attributes

jeremiedbb committed:
fix predict test and use assert_allclose

jeremiedbb committed:
Merge remote-tracking branch 'upstream/main' into pr/marenwestermann/23934

jeremiedbb committed:
check [all random seeds]
test_n_samples_leaves_roots
test_partial_fit
test_birch_predict
test_n_clusters
test_sparse_X
test_branching_factor
test_threshold
test_transform_match_across_dtypes

jeremiedbb committed:
check [all random seeds]
test_n_samples_leaves_roots
test_partial_fit
test_birch_predict
test_n_clusters
test_sparse_X
test_branching_factor
test_threshold
test_transform_match_across_dtypes

jeremiedbb reviewed (approved):
Thanks for the PR @marenwestermann. LGTM
